Objects/genericaliasobject.c

```diff
@@ -644,10 +647,10 @@ Py_GenericAlias(PyObject *origin, PyObject *args)
     if (alias == NULL) {
         return NULL;
     }
+    _PyObject_GC_TRACK(alias);
     if (!setup_ga(alias, origin, args)) {
-        PyObject_GC_Del((PyObject *)alias);
```

Comment:
I dislike the idea of having a code different in 3.9 and main. Why would 3.9 have different code?

It looks like a legit bug which also affects the main branch.

Reply:
I dislike tracking the object earlier, before it's fully initialized. It goes against https://bugs.python.org/issue44531

I would prefer to track the object if setup_ga() fails, so ga_dealloc() doesn't have to use PyObject_GC_UnTrack() (to not untrack if it is not track).

An alternative is to use PyObject_GC_UnTrack() in ga_dealloc(), only for this code path.

@pablogsal: I also have this issue in https://bugs.python.org/issue44531 Is it a good practice to track an object partially initialized (its traverse function may crash) only to be able to deallocate it?

I didn't check if ga_traverse() can be called (don't crash) if setup_ga() fails. In your PR, you change the function.

Reply:
The code is also like this on the main branch: https://github.com/python/cpython/blob/main/Objects/genericaliasobject.c#L655

This is a backport PR. The only part I changed was this line onwards https://github.com/python/cpython/pull/27028/files#diff-828d12085a29364c67442b193bb62906e3469fbe21367499fd62817f98190014R36.

I'm waiting for Pablo to merge #27021 first. This PR is just to test if I can fix the failures on macOS, because I can't repro them on main or 3.10.

Reply:
Sorry if I wasn't very clear, a quick summary:
- Calling ``PyObject_GC_Del`` is apparently not good and caused buildbots to fail for ``types.Union`` https://github.com/python/cpython/pull/27008#discussion_r663398434.
- So after fixing ``types.Union``, I created PR #27016 for ``types.GenericAlias`` because it has the same pattern.
- For some reason, macOS buildbot passed on main and 3.10, but failed on 3.9, even though I don't see any significant code change in ``genericaliasobject.c`` and ``gcmodule.c``
- We reverted that backport 3.9 PR, but now I think I found the issue so I'm resubmitting it. The failures may be caused by ``Py_VISIT`` being called on ``NULL`` ``alias->parameters`` in ``ga_traverse``.

Reply:
Hmm, the macOS test is still failing, I think you're right here. It's probably triggering GC on improperly initialized object :(.

Reply:
> @pablogsal: I also have this issue in https://bugs.python.org/issue44531 Is it a good practice to track an object partially initialized (its traverse function may crash) only to be able to deallocate it?

If you use `PyGenericAlloc`, the fields are initialized to NULL and the `VISIT` macro handles that correctly